src/toast/io/observation_hdf_save_v1.py

```diff
@@ -0,0 +1,984 @@
+                        del recv_buf
+
+                        hf_bytes += n_recv_bytes
+                        hf_det += n_recv_dets
+
+                elif proc == rank:
+                    # We are sending.  First send the number of bytes and detectors
+                    det_ranges = np.zeros(
+                        (len(comp_ranges), 2),
+                        dtype=np.int64,
+                    )
+                    for d in range(len(comp_ranges)):
+                        det_ranges[d, :] = comp_ranges[d]
+                    comm.send(n_local_bytes, dest=0, tag=tag_offset + 10 * proc)
+                    comm.send(len(det_ranges), dest=0, tag=tag_offset + 10 * proc + 1)
+
+                    comm.Send(comp_bytes, dest=0, tag=tag_offset + 10 * proc + 2)
+                    comm.Send(
+                        det_ranges.flatten(), dest=0, tag=tag_offset + 10 * proc + 3
+                    )
+                    if comp_data_offsets is not None:
+                        comm.Send(
+                            comp_data_offsets, dest=0, tag=tag_offset + 10 * proc + 4
+                        )
+                    if comp_data_gains is not None:
+                        comm.Send(
+                            comp_data_gains, dest=0, tag=tag_offset + 10 * proc + 5
+                        )
+
+
+@function_timer
+def save_hdf5_intervals(obs, hgrp, fields, log_prefix):
+    log = Logger.get()
+
+    timer = Timer()
+    timer.start()
+
+    # We are using the group communicator
+    comm = obs.comm.comm_group
+    nproc = obs.comm.group_size
+    rank = obs.comm.group_rank
+
+    for field in fields:
+        if field not in obs.intervals:
+            msg = f"Intervals '{field}' does not exist in observation "
+            msg += f"{obs.name}.  Skipping."
+            log.warning_rank(msg, comm=comm)
+            continue
+
+        if field == obs.intervals.all_name:
+            # This is the internal fake interval for all samples.  We don't
+            # save this because it is re-created on demand.
+            continue
+
+        # Get the list of start / stop tuples on the rank zero process
+        ilist = global_interval_times(obs.dist, obs.intervals, field, join=False)
+
+        n_list = None
+        if rank == 0:
+            n_list = len(ilist)
+        if comm is not None:
+            n_list = comm.bcast(n_list, root=0)
+
+        # Participating processes create the dataset
+        hdata = None
+        if hgrp is not None:
+            hdata = hgrp.create_dataset(field, (2, n_list), dtype=np.float64)
+            # Only the root process writes
+            if rank == 0:
+                hdata[:, :] = np.transpose(np.array(ilist))
+        del hdata
+
+        log.verbose_rank(
+            f"{log_prefix}  Intervals finished {field} write in",
+            comm=comm,
+            timer=timer,
+        )
+
+
+def save_instrument(parent_group, telescope, comm=None, session=None):
```

Comment:
Is this essentially implementing the `save_hdf5()` method for the instrument class? Where do we draw the line between classes that implement `save_hdf5()` and classes that have it implemented externally like this?

Reply:
That is a *really* good point.  We ended up in this situation because previously we were not saving/loading a single object here (so we had to do it manually).  But now we are writing the `telescope` and all its members to an HDF5 group.  So you are correct that we should move these functions to methods of the `Telescope` class.

Reply:
This is paradigm is now implemented.  All instrument classes implement save / load methods and the observation save / load makes use of those.